ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PlanNode.java

```diff
@@ -42,7 +41,6 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-@Immutable
```

Comment:
Because the note is already part of the logical plan. So it would need to swap itself out, which would mean the parent nodes reference would need to change to the node. Ergo, the parent node would need to be mutable.

The general direction is to move more and more logic from the `Analyzer` into the logical plan nodes, as this allows us to make use of the plan hierarchy and polymorphism to drive decision making. So more and more, the logical plan nodes, as they are today, will become more mutable.

For example, the final user-provided projection can be used to limit the number of columns we need to extract from each source.  The final projection node can inform parent nodes of the required columns. This would allow us to, for example, limit the columns we repartition to only those needed down stream, not the whole message: this could potentially be a massive saving.

If we want an immutable logical model, then we'll need a two-step process, i.e. a set of builder nodes, that can be baked into a immutable model. 

I'm a big fan of immutability, especially in the presence of multi-threading. However, in this case, I'm not sure it actually buys us much, at least not at the moment.